py-rattler/src/networking/middleware.rs

```diff
@@ -0,0 +1,70 @@
+use pyo3::{pyclass, pymethods, FromPyObject};
+use rattler_networking::{
+    mirror_middleware::Mirror, AuthenticationMiddleware, AuthenticationStorage, MirrorMiddleware,
+};
+use std::collections::HashMap;
+use url::Url;
+
+#[derive(FromPyObject)]
+pub enum PyMiddleware {
+    MirrorMiddleware(PyMirrorMiddleware),
+    AuthenticationMiddleware(PyAuthenticationMiddleware),
+}
+
+#[pyclass]
+#[repr(transparent)]
+#[derive(Clone)]
+pub struct PyMirrorMiddleware {
+    pub(crate) inner: HashMap<Url, Vec<Mirror>>,
+}
+
+#[pymethods]
+impl PyMirrorMiddleware {
+    #[new]
+    pub fn __init__(inner: HashMap<String, Vec<String>>) -> Self {
+        let map = inner
+            .into_iter()
+            .map(|(k, v)| {
+                (
+                    Url::parse(&k).unwrap(),
+                    v.into_iter()
+                        .map(|url| Mirror {
+                            url: Url::parse(&url).unwrap(),
```

Comment:
Can we confidently unwrap here?

Reply:
[`e8c631c` (#915)](https://github.com/conda/rattler/pull/915/commits/e8c631c2118c3c8be9df05b13b8ba107d9e552a2)